THREE.Box3 getCorners method added

Wilt opened the pull request:
I ran often into the case that I need the corners of a bounding box (`THREE.Box3`) as vertices.

In this PR a `getPoints` method was added to `THREE.Box3`. At first I also thought of refactoring the applyMatrix method so it uses the `getPoints` but I realized that the points variable inside the wrapper is there probably for performance reasons so that there is no new array with points created for each time `applyMatrix` is called. So I left the original method.

The method could also be called `getCorners`.

Commit:
Added getPoints method to Box3

Commit:
Added getPoints method to Box3

Commit:
Added getPoints method to Box3

WestLangley commented:
If @mrdoob and @bhouston feel this method is appropriate, I would suggest, at a minimum, that a second argument `optionalTarget` be added to avoid instantiating a new array with each call. Follow the coding pattern used in the other math classes.

Perhaps there is a better method name.

`Box2/3.getVertices()` ?

`Box2/3.toArray()` ?

bhouston commented:
I am fine with an optional target and then we can use it here, instead of having this list of points duplicated:

https://github.com/Wilt/three.js/blob/dad13a4c1cca77a776da85807dfebe2cfb12bdeb/src/math/Box3.js#L436

Also I'd call it getCornerPositions() or getCorners() -- something along those lines, more descriptive.

bhouston commented:
I think that ```toCorners()``` is a nice name and a combo of what has been suggested.

Wilt commented:
I like `getCorners` is also used [in other libraries](https://msdn.microsoft.com/en-us/library/bb197066.aspx)...

Commit:
Renamed method to getCorners
Added optional target so target can be passed to method
Used getCorners method inside applyMatrix and passed points as target

Added unit tests for getCorners method

Wilt commented:
- Renamed `getPoints` method to `getCorners`
- Added `optionalTarget` param to `getCorners` so target can be passed when desired
- Used `getCorners` method inside `applyMatrix` and passed `points` variable as target
- Added unit tests for `getCorners` method for both non target and target user case

bhouston commented on src/math/Box3.js:
```diff
@@ -399,33 +399,60 @@ THREE.Box3.prototype = {
+
 	applyMatrix4: function () {
 
 		var points = [
```
Try this in the closure:
```
var points = null;
```

And then lower down use this pattern:
```
points = this.getCorners( points );
```

And now it will allocate the first time it is used.  Less code and lazy allocation. :)

Commit:
Initialize points on first call, less code and lazy allocation according to @bhouston his suggestion

Wilt commented:
@bhouston Nice, thanks for the suggestion, I updated accordingly...

bhouston commented:
This PR looks great to me.

WestLangley commented:
@Wilt Can you update the docs, too?

@bhouston Maybe add the new method to `Box2` -- or wait. What do you think?

Commit:
Added documentation. Someone should double check before merge...

Wilt commented:
@WestLangley Docs are updated. Someone should double check before merge. I wasn't sure how to doc the return value (array of vertices) correctly. I did `THREE.Vector3[]`. But it seems the return values are not used in all classes and couldn't find another example of an array being returned.

I also noticed return values are not always correct; almost everywhere is written `this` even when it should be something else. For example [here](http://threejs.org/docs/#Reference/Math/Box3.size). Maybe I should make another PR for that, but then I need some input on how to document return values correctly...

Wilt commented:
I also wanted to add a test, but I had issues getting the test running. I needed to include a lot of extra files (load test dependencies) for making the test run correctly. To test I needed Mesh, Object3D, Geometry, BufferGeometry, etc etc. I stopped with that because I am not sure if these kind of dependencies are OK for testing Math classes. I could continue with this on request...

Commit:
Added documentation. Someone should double check before merge...

Wilt commented:
Before this is merged maybe we should also discuss the order of the points.

Now it is like this:
````
new THREE.Vector3(), // 000
new THREE.Vector3(), // 001
new THREE.Vector3(), // 010
new THREE.Vector3(), // 011
new THREE.Vector3(), // 100
new THREE.Vector3(), // 101
new THREE.Vector3(), // 110
new THREE.Vector3()  // 111

   3____7
 2/|__6/|
 | 1__|_5
 0/___4/

````
Maybe would be better:

````
   7____6
 3/|__2/|
 | 4__|_5
 0/___1/
````

bhouston commented:
Do we want to make an ordering guarantee in the docs?  If so feel free to pick an appropriate order.  I am not attached to the order I coded up originally in this function because I didn't expose it outside of the applyMatrix function.

I do always have a biased to doing what other libraries are doing, unless there is a strong argument against -- so if you can find an ordering another popular library is using, I'd prefer that ordering, whatever it is.

Wilt commented:
[Here](https://msdn.microsoft.com/en-us/library/bb197066.aspx) they do:

> The points returned correspond to the corners of the BoundingBox faces that are perpendicular to the z-axis. The near face is the face with the larger z value, and the far face is the face with the smaller z value. Points 0 to 3 correspond to the near face in a clockwise order starting at its upper-left corner when looking toward the origin from the positive z direction. Points 4 to 7 correspond to the far face in a clockwise order starting at its upper-left corner when looking toward the origin from the positive z direction.

So that would mean:
````
min.x, max.y, max.z // 011
max.x, max.y, max.z // 111
max.x, min.y, max.z // 101
min.x, min.y, max.z // 001
min.x, max.y, min.z // 010
max.x, max.y, min.z // 110
max.x, min.y, min.z // 100
min.x, min.y, min.z // 000

  0____1
4/|__5/|   y
| 3__|_2   | z
7/___6/    0/___x
````

bhouston commented:
DirectX is a really good model to follow.  I support this change unless you have a reason to pick a different ordering....

Commit:
Updated with new order of points according to DirectX bounding box model

Wilt commented:
Updated according to @bhouston his suggestion...

Wilt commented:
@mrdoob Any chances of merging this?

Wilt commented:
I would like to bump this one. Still would like this to be merged...

WestLangley commented:
@Wilt Can you remove the conflicts? Closures are implemented differently, now.

Commit:
merged with upstream dev

Commit:
merged with upstream dev

Wilt commented:
@WestLangley So i managed to solve the conflicts, but now I was unable to rerun my tests. Before I was able to run them but seems I cannot get the new QUnit test environment setup correctly.

I will try to find some more time later to continue working on this.
Otherwise you guys can maybe run the tests when merging (I assume you do that anyway).

WestLangley commented on src/math/Box3.js:
```diff
@@ -428,41 +428,67 @@ Object.assign( Box3.prototype, {
+         * 4/|__5/|   y
+         * | 3__|_2   | z
+         * 7/___6/    0/___x
+         *
```
This is a bit confusing. three.js uses a right-handed coordinate system; this illustration is left-handed.

Wilt commented:
@WestLangley OK, thanks for your last comment. 
My axes are actually corresponding to the right hand coordinate system, but the z is pointing up. As having a background in the building industry I am used to having z pointing up. I can revise this. What is more interesting is the comment made me double check the handedness of DirectX. It turns out DirectX uses theleft hand coordinate system. I missed that before. That means that translating this:

> The points returned correspond to the corners of the BoundingBox faces that are perpendicular to the z-axis. The near face is the face with the larger z value, and the far face is the face with the smaller z value. Points 0 to 3 correspond to the near face in a clockwise order starting at its upper-left corner when looking toward the origin from the positive z direction. Points 4 to 7 correspond to the far face in a clockwise order starting at its upper-left corner when looking toward the origin from the positive z direction.

Makes probably no sense in a coordinate system with a different handedness right?

Can @WestLangley or @bhouston still give some feedback on this, then I can still change if needed.
I would suggest using the following order of points which is pretty intuitive:

```
min.x, min.y, min.z // 000
max.x, min.y, min.z // 100
max.x, max.y, min.z // 110
min.x, max.y, min.z // 010
min.x, min.y, max.z // 001
max.x, min.y, max.z // 101
max.x, max.y, max.z // 111
min.x, max.y, max.z // 011

  7____6
3/|__2/|   y
| 4__|_5   | z
0/___1/    0/___x

  7____6
3/|__2/|   y
| 4__|_5   | z
0/___1/    0/___x
```

WestLangley commented:
Here is the three.js right-handed coordinate system as it is normally presented with the camera located at ( 1, 1, 1 ).
```
// right-handed y up
  7____6    y
3/|__2/|    |
| 4__|_5    0 ___ x
0/___1/    /
         z
```
I do not care what the order is. The order in the above illustration is fine.

Commit:
Updated points since DirectX is using different handedness, it doesn't make sense to use that order.
Removed confusing comment

Mugen87 commented:
Closing due to merge conflicts and because optional targets are now mandatory. For consistency reasons, it would be also necessary to add the method for `Box2`.

Apart from that, I'm open for this addition since I've seen similar methods in other AABB classes e.g. 

https://github.com/juj/MathGeoLib/blob/db162d35bd704e8946508b0e0cd815a2c6934252/src/Geometry/AABB.h#L241-L244
